mlflow/sklearn/__init__.py

```diff
@@ -455,6 +460,323 @@ def load_model(model_uri):
+        So that in following metric autologging, the metric will be logged into the registered
+        run_id
+        """
+        model._mlflow_run_id = run_id
+
+    @staticmethod
+    def gen_name_with_index(name, index):
+        assert index >= 0
+        if index == 0:
+            return name
+        else:
+            # Use '-' as the separator between name and index,
+            # The '-' is not valid character in python var name
+            # so it can prevent name conflicts after appending index.
+            return f'{name}-{index + 1}'
+
+    def register_eval_dataset(self, model, eval_dataset):
+        """
+        Register eval dataset into eval_dataset_info_map, it will do:
+         1. inspect eval dataset var name.
+         2. check whether eval_dataset_info_map already registered this eval dataset.
+            will check by object id and comparing row samples.
+            Note: only check object id is not sufficient, id may be reused.
+         3. register eval dataset (with id and row samples information)
+         4. return eval dataset name with index.
+
+        Note: this method include inspecting argument variable name.
+         So should be called directly from the "patched method", to ensure it capture
+         correct argument variable name.
+        """
+        eval_dataset_name = _inspect_original_var_name(
+            eval_dataset, fallback_name='unknown_dataset')
+        eval_dataset_id = id(eval_dataset)
+
+        run_id = self.get_run_id_for_model(model)
+        registered_dataset_list = self._eval_dataset_info_map[run_id][eval_dataset_name]
+
+        index = len(registered_dataset_list)
+        for i, id_i in enumerate(registered_dataset_list):
+            if eval_dataset_id == id_i:
+                index = i
+                break
+
+        if index == len(registered_dataset_list):
+            # register new eval dataset
+            registered_dataset_list.append(eval_dataset_id)
+
+        return self.gen_name_with_index(eval_dataset_name, index)
+
+    def register_prediction_result(self, model, eval_dataset_name, predict_result):
+        """
+        In `patched_predict`, register the prediction result instance with the model id and eval
+        dataset name. e.g.
+
+        ```
+        prediction_result = model_1.predict(eval_X)
+        ```
+        then we need register an item into the `pred_result_id_to_dataset_name_and_run_id` map like
+        below:
+        id(prediction_result) --> (inspected_original_var_name_of(eval_X), id(model_1))
+
+        With this map `dataset_id_to_dataset_name_and_run_id`, in following patched metric API,
+        we can query the eval dataset name and the model id via the `y_pred` argument instance.
+        """
+        value = (eval_dataset_name, self.get_run_id_for_model(model))
+        prediction_result_id = id(predict_result)
+        self._pred_result_id_to_dataset_name_and_run_id[prediction_result_id] = value
```

Comment:
Nit: does `value` need to be defined?

```suggestion
        prediction_result_id = id(predict_result)
        self._pred_result_id_to_dataset_name_and_run_id[prediction_result_id] = (
            eval_dataset_name, self.get_run_id_for_model(model)
        )
```